### PR TITLE
Overlapping AOI Polygon Check

### PR DESCRIPTION
**Pull request by Bentleysb**
This check flags polygons that overlap and represent the same Area of Interest (AOI).

AOIs are defined through the configurable value `aoi.tags.filters`. This is a list of tag filters.  
If an Area has any of the tags in any of the filters, it is considered an AOI.  
If two overlapping AOIs both have tags in the same filter, they are flagged as representing the same thing.

**OSM Examples:**

1. Way [id:23182647](https://www.openstreetmap.org/way/23182647) overlaps way [id:301044293](https://www.openstreetmap.org/way/301044293) and they are both tagged as `landuse=cemetery`.
2. Way [id:203547456](https://www.openstreetmap.org/way/203547456), tagged as `natural=wood`, overlaps way [id:546230157](https://www.openstreetmap.org/way/546230157), tagged as `landuse=forest`, both tags are representing the same AOI.

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+        final Area aoi = (Area) object;
+        final Polygon aoiPolygon = aoi.asPolygon();
+        boolean hasOverlap = false;
+
+        // Set of overlapping area AOIs
+        final Set<Area> overlappingArea = Iterables.stream(object.getAtlas().areasIntersecting(
+                aoiPolygon.bounds(),
+                area -> area.getIdentifier() != aoi.getIdentifier() && area.intersects(aoiPolygon)
+                        && aoiFiltersTest(area) && !this.isFlagged(area.getIdentifier())))
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: It might be worthwhile moving the `isFlagged` check before the intersects check.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+
+        final Area aoi = (Area) object;
+        final Polygon aoiPolygon = aoi.asPolygon();
+        boolean hasOverlap = false;
+
+        // Set of overlapping area AOIs
+        final Set<Area> overlappingArea = Iterables.stream(object.getAtlas().areasIntersecting(
+                aoiPolygon.bounds(),
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: `bounds() `has the potential to be an expensive call, consider pulling it out of the stream. I'd be curious to see if you get any performance increase with that.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+
+        final Area aoi = (Area) object;
+        final Polygon aoiPolygon = aoi.asPolygon();
+        boolean hasOverlap = false;
+
+        // Set of overlapping area AOIs
+        final Set<Area> overlappingArea = Iterables.stream(object.getAtlas().areasIntersecting(
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: Nit: overlappingArea --> overlappingAreas

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+     * This is the actual function that will check to see whether the object needs to be flagged.
+     *
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return an optional {@link CheckFlag} object that
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: Nit: extra line

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        final CheckFlag flag = new CheckFlag(this.getTaskIdentifier(object));
+        flag.addObject(object);
+
+        // Test each overlapping AOI to see if it overlaps enough and passes the same AOI filter as
+        // the object
+        for (final Area area : overlappingArea)
+        {
+            if (this.hasMinimumOverlapProportion(aoi.asPolygon(), area.asPolygon()))
+            {
+                if (aoiFiltersTest(object, area))
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: Consider combining the if statements to improve readability and reduce indentation.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    /**
+     * Uses code from findIntersectionType function in BuildingRoadIntersection
+     * 
+     * @param polygon
+     *            {@link Polygon} to check for intersection
+     * @param otherPolygon
+     *            Another {@link Polygon} to check against for intersection
+     * @return true if the polygon to polygon overlap surface area is greater than minimum
+     */
+
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: Nit: extra line

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    private boolean hasMinimumOverlapProportion(final Polygon polygon, final Polygon otherPolygon)
+    {
+
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: Nit: extra line

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    }
+
+    @Override
+    protected List<String> getFallbackInstructions()
+    {
+        return FALLBACK_INSTRUCTIONS;
+    }
+
+    /**
+     * Uses code from findIntersectionType function in BuildingRoadIntersection
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: If there is enough reuse, can you pull out the common functionality into an intersection utility or have a base class for all intersection checks?

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+    {
+
+        Clip clip = null;
+        try
+        {
+            clip = polygon.clip(otherPolygon, Clip.ClipType.AND);
+        }
+        catch (final TopologyException e)
+        {
+            System.out
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: This should be replaced with logger. How often does this happen?

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+import com.vividsolutions.jts.geom.TopologyException;
+
+/**
+ * This check flags {@link Area}s that overlap and represent the same Area of Interest (AOI) AOIs
+ * are defined by {@link TaggableFilter}s, which are grouped in a {@link List}.
+ *
+ * @author danielbaah
+ * @author bbreithaupt
+ */
+public class OverlappingAOIPolygonCheck extends BaseCheck
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: Really high-level question - are there AOIs represented by Relations? If so, we want some kind of ComplexEntity implementation to walk the relation and build a valid Polygon that represents the AOI and update this check for completeness. If this is the case, can you please add an issue to the Atlas project for such functionality? Or you can confirm here, and I can make the issue :)

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    /**
+     * Tests the input {@link AtlasObject} against the list of {@link TaggableFilter}s.
+     *
+     * @param object
+     *            {@link AtlasObject} to be tested
+     * @return true if any of the filters return true
+     */
+    private boolean aoiFiltersTest(final AtlasObject object)
+    {
+        for (final TaggableFilter filter : this.aoiFilters)
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: Nit: consider stream approach here to simplify: `return this.aoiFilters.stream().anyMatch(filter -> filter.test(object));`

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+     *
+     * @param object
+     *            {@link AtlasObject} to be tested
+     * @param area
+     *            {@link Area} to be tested against {@code object}
+     * @return true if any of the filters return true for both {@code object} and {@code area}
+     */
+    private boolean aoiFiltersTest(final AtlasObject object, final Area area)
+    {
+        for (final TaggableFilter filter : this.aoiFilters)
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: Same comment about `anyMatch` functionality.

File: src/test/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheckTest.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+import org.openstreetmap.atlas.checks.configuration.ConfigurationResolver;
+import org.openstreetmap.atlas.checks.validation.verifier.ConsumerBasedExpectedCheckVerifier;
+import org.openstreetmap.atlas.utilities.configuration.Configuration;
+
+/**
+ * Tests for {@link OverlappingAOIPolygonCheck}
+ *
+ * @author bbreithaupt
+ */
+public class OverlappingAOIPolygonCheckTest
```
Comment (MikeGost) [inline src/test/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheckTest.java]: Does this test all possible code paths - overlap and non-overlap, overlapping but not matching filters, overlapping but not having enough overlap?

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java (hunk 2 above)
Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: Nice catch, there is a performance increase of about 0.1%.

File: src/test/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheckTest.java (hunk 13 above)
Comment (Bentleysb) [reply src/test/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheckTest.java]: It was missing a test for overlap that is below the minimum. That is now present.

File: docs/checks/overlappingAOIPolygonCheck.md [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+
+        final Area aoi = (Area) object;
+        final Polygon aoiPolygon = aoi.asPolygon();
+        boolean hasOverlap = false;
+
+        // Set of overlapping area AOIs
+        final Set<Area> overlappingArea = Iterables.stream(object.getAtlas().areasIntersecting(
+                aoiPolygon.bounds(),
```
Comment (MikeGost) [inline docs/checks/overlappingAOIPolygonCheck.md]: I'm not too familiar with how this is setup. Two questions:
1. Why is the java code duplicated in the markdown document?
2. Should it be consistent with the changes made to the code below?

Comment (danielduhh) [reply docs/checks/overlappingAOIPolygonCheck.md]: Great question; the markdown docs are supposed to be a high level walkthrough, or tutorial, of the check code. This includes flag examples and a walkthrough, as well as detailed explanation of any complex/interesting functions.

In this case, there's probably more code copied over than needed 😃

Comment (Bentleysb) [reply docs/checks/overlappingAOIPolygonCheck.md]: The markdown document is meant to give a higher level explanation of what the code is doing. Explanations of the code are paired with the code snippets to give context, that is why there is the duplication. As for the consistency, it should be the same. I will correct that.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java (hunk 8 above)
Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: I am looking into this.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java (hunk 10 above)
Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: I am not sure. I will check and report back.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java (hunk 8 above)
Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: I have moved this to a new utility class. Once this PR is merged I will look into updating IntersectingBuildingsCheck to use that class as well. 

IntersectingBuildingsCheck was were the code was originally from. The java doc in this check incorrectly listed it as from BuildingRoadIntersection.

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java (hunk 9 above)
Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: This appears to occur at a rate of about once per 30 flags.

File: src/main/java/org/openstreetmap/atlas/checks/validation/utility/IntersectionUtilities.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
+package org.openstreetmap.atlas.checks.validation.utility;
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/checks/validation/utility/IntersectionUtilities.java]: If you're going the global utility route, it would make more sense to put the utility package under `org.openstreetmap.atlas.checks`

File: docs/checks/overlappingAOIPolygonCheck.md (hunk 14 above)
Comment (MikeGost) [reply docs/checks/overlappingAOIPolygonCheck.md]: That makes sense - do we want to continue doing this? It sounds like the two can quickly get out of sync - as they are in this PR already. And looking at the last few, it's been the same code as in the actual implementation. It may be worthwhile updating the strategy around this. Just my 2 cents, you have more insight in this project!

File: src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java (hunk 10 above)
Comment (MikeGost) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: Cool - let me know what you find, might be an edge case.

Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: It does appear that AOIs are commonly represented by relations. If you are still up for putting in that Atlas issue, I think it might be best for you to do it as I am not particularly familiar with ComplexEntity.

Comment (MikeGost) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: Sure, I can create the issue and we can update this check when it's ready (not in the near future). Can you provide some of the examples you found?

Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: A relations representing an AOI:
https://www.openstreetmap.org/relation/2985897

A relation representing an AOI overlapping a way representing an AOI:
https://www.openstreetmap.org/relation/1279007
https://www.openstreetmap.org/way/583819702

Comment (MikeGost) [reply src/main/java/org/openstreetmap/atlas/checks/validation/areas/OverlappingAOIPolygonCheck.java]: Thanks! These are great examples. I've created the issue here: https://github.com/osmlab/atlas/issues/163. Once that's implemented, we can come back and update this check to capture more use cases.

**Comment by MikeGost** [comment]
This looks ready to go. Leaving for @jklamer to review.
